File: src/shader.c
#include "shader.h"
#include <unistd.h>

#include <stdio.h>
#include <string.h>

#include "glad/gl.h"
#include "cglm/cglm.h"

u32 shader_load(u32 type, string_s path) {
    char cwd[1024];
    getcwd(cwd, sizeof(cwd));
    cwd[strlen(cwd) - 17] = '\0';

    char abspath[1024];

    snprintf(abspath, sizeof(abspath), "%s%s", cwd, path.data);

    FILE* file = fopen(abspath, "rb");
    if (file == NULL) {
        fprintf(stderr, "Could not open %s\n", abspath);
        exit(-1);
    }

    fseek(file, 0, SEEK_END);
    usize file_size = ftell(file);
    rewind(file);

    char* buffer = malloc(file_size + 1);
    fread(buffer, sizeof(char), file_size, file);
    buffer[file_size] = '\0';
    fclose(file);


    u32 shader = glCreateShader(type);
    glShaderSource(shader, 1, (const char**) &buffer, NULL);
    glCompileShader(shader);

    b32 success;
    glGetShaderiv(shader, GL_COMPILE_STATUS, &success);
    if (!success) {
        char info_log[1024];
        glGetShaderInfoLog(shader, 1024, NULL, info_log);
        fprintf(stderr, "Could not compile %s\nError: %s", abspath, info_log);
    }

    free(buffer);

    return shader;
}

shader_s shader_create(string_s vpath, string_s fpath) {
    shader_s shader = {0};

    u32 vshader = shader_load(GL_VERTEX_SHADER, vpath);
    u32 fshader = shader_load(GL_FRAGMENT_SHADER, fpath);

    shader.id = glCreateProgram();
    glAttachShader(shader.id, vshader);
    glAttachShader(shader.id, fshader);
    glLinkProgram(shader.id);


    b32 success;
    glGetProgramiv(shader.id, GL_LINK_STATUS, &success);
    if (!success) {
        char info_log[1024];
        glGetProgramInfoLog(shader.id, 1024, NULL, info_log);
        fprintf(stderr, "Could not link %d\nError: %s", shader.id, info_log);
    }

    glDeleteShader(vshader);
    glDeleteShader(fshader);

    return shader;
}

void shader_bind(const shader_s* shader) {
    glUseProgram(shader->id);
}

void shader_unbind() {
    glUseProgram(0);
}

void shader_set_matrix(const shader_s* shader, string_s name, const mat4* matrix) {
    glUniformMatrix4fv(glGetUniformLocation(shader->id, name.data), 1, false, (f32*) matrix);
}

void shader_destroy(shader_s* shader) {
    shader_unbind();
    glDeleteProgram(shader->id);
}

File: src/Shader.zig
const std = @import("std");
const c = @import("c.zig").c;
const zalgebra = @import("zalgebra");

const Error = error{ ShaderCompilationFailed, ShaderLinkingFailed };

const Shader = @This();

handle: u32,

pub fn init(allocator: std.mem.Allocator, io: std.Io, vertex_path: [:0]const u8, fragment_path: [:0]const u8) !Shader {
    const vertex_shader = try createShader(io, allocator, vertex_path, c.GL_VERTEX_SHADER);
    const fragment_shader = try createShader(io, allocator, fragment_path, c.GL_FRAGMENT_SHADER);

    const handle: u32 = c.glCreateProgram();
    c.glAttachShader(handle, vertex_shader);
    c.glAttachShader(handle, fragment_shader);
    c.glLinkProgram(handle);

    var success: i32 = 0;
    c.glGetProgramiv(handle, c.GL_LINK_STATUS, @ptrCast(&success));

    if (success != c.GL_TRUE) {
        var buffer: [1024]u8 = undefined;
        var info_len: u32 = undefined;

        c.glGetShaderiv(handle, c.GL_INFO_LOG_LENGTH, @ptrCast(&info_len));
        c.glGetShaderInfoLog(handle, buffer.len, @ptrCast(&info_len), &buffer);

        std.log.err("Shader linking failed! Error: {s}", .{buffer[0..info_len]});

        return Error.ShaderLinkingFailed;
    }

    return .{
        .handle = handle,
    };
}

pub fn deinit(self: *Shader) void {
    c.glDeleteProgram(self.handle);
}

pub fn bind(self: Shader) void {
    c.glUseProgram(self.handle);
}

pub fn setMat4(self: Shader, mat4: zalgebra.Mat4x4(f32), name: [:0]const u8) void {
    c.glUniformMatrix4fv(c.glGetUniformLocation(self.handle, name), 1, c.GL_FALSE, @ptrCast(&mat4.data[0][0]));
}

fn createShader(io: std.Io, allocator: std.mem.Allocator, path: []const u8, shader_type: u32) !u32 {
    const src = try allocator.dupeSentinel(u8, try std.Io.Dir.cwd().readFileAlloc(io, path, allocator, .limited(std.math.maxInt(usize))), 0);
    defer allocator.free(src);

    const shader = c.glCreateShader(shader_type);

    c.glShaderSource(shader, 1, @ptrCast(&src), null);
    c.glCompileShader(shader);

    var success: i32 = 0;
    c.glGetShaderiv(shader, c.GL_COMPILE_STATUS, @ptrCast(&success));

    if (success != c.GL_TRUE) {
        var buffer: [1024]u8 = undefined;
        var info_len: u32 = undefined;

        c.glGetShaderiv(shader, c.GL_INFO_LOG_LENGTH, @ptrCast(&info_len));
        c.glGetShaderInfoLog(shader, buffer.len, @ptrCast(&info_len), &buffer);

        std.log.err("Shader compilation failed! Error: {s}", .{buffer[0..info_len]});

        return Error.ShaderCompilationFailed;
    }

    return shader;
}
